enhanced-stack-trace-queries.md

```diff
@@ -0,0 +1,84 @@
+# Enhanced stack trace queries
+
+## Backround
+
+The SDKs have a built-in `__stack_trace` query that can be used display where the workflow is blocked at the time of the
+query (the last processed workflow task).
+
+The response to a `__stack_trace` query is a single string representing a list of stack individual stack traces
+separated by double newlines.
+
+Example from the TS SDK:
+
+```
+    at someInternalFunction (/path/to/workflows/my-workflow.ts:18:10)
+    at someWorkflowFunction (/path/to/workflows/my-workflow.ts:60:20)
+
+    at someInternalFunction2 (/path/to/workflows/my-workflow.ts:100:30)
+    at someWorkflowFunction2 (/path/to/workflows/my-workflow.ts:123:5)
+```
+
+While this is valuable for developers to be able to understand how their code correlates to event histories, it is quite
+limited since it requires looking up the actual code to get the full picture.
+
+## Proposal
+
+SDKs can optionally implement an `__enhanced_stack_trace` query (name TBD) where instead of returning the stack traces
+as a single string, they return a pre-defined structure, that structure will also include file sources.
+
+The web UI can use this response to show the workflow code and highlight locations where it is blocked.
+
+### Proposed structure
+
+```ts
+/**
+ * A pointer to a location in a file
+ */
+export interface FileLocation {
+  filename: string;
+  line: number;
+  column: number;
+}
+
+/**
+ * Used as the result for the enhanced stack trace query
+ */
+export interface EnhancedStackTrace {
+  /**
+   * Mapping of file name to file contents
+   */
+  sources: Record<string, string>;
```

Comment:
A few things to consider here:
- This would only contain files that are mentioned in actual stack traces, right?
- Would there be any value including sources for the SDK itself, or does it make sense to exclude them?
- Should we worry about some limit on maximum query response length? AFAIK query responses are subject to grpc's max packet size of 4 mb, but is there some other, lower limit that could be an issue?

Reply:
> A few things to consider here:
> 
> * This would only contain files that are mentioned in actual stack traces, right?

Yes

> * Would there be any value including sources for the SDK itself, or does it make sense to exclude them?

We typically cut the stack trace at the SDK level in the stack trace query so those should only make it to the UI in rare situations, we'll probably 

> * Should we worry about some limit on maximum query response length? AFAIK query responses are subject to grpc's max packet size of 4 mb, but is there some other, lower limit that could be an issue?

That's a concern for sure, one option for dealing with this is to limit the source code sent to X wrapping lines for each location but I'm not sure it's worth the optimization here, worth considering.

Reply:
Forget this. If we agree on trimming down the list of files being transferred to only _user_ files that are indeed mentioned in the stack traces, then I don't think this is an immediate issue. We might have to deal with some huge source files at some point in the future, but that can be dealt later.

Reply:
Even if just the first user source file, could be large. More concerned with unnecessary data/load time than avoiding reaching max possible. Would do whatever we wanted to display, like 15 lines on either side, to reduce load time.

Reply:
>  like 15 lines on either side

Not that easy given the proposed structure... We might have several references to a same file, at different locations. We would have to capture 15 lines on either side starting from each active position in that file... And then, we end up with a sparse file... How can the UI match line numbers between the stack trace and the sparse source file?

Reply:
Could do:

```
sources: Record<Filename, Lines>;

Lines = Record<number, string>;
```

Reply:
Or better yet:
```
sources: Record<string, ContentWithOffset>

interface ContentWithOffset {
  content: string;
  lineOffset: number;
}
```

Reply:
What's an example?

Reply:
What do you mean?

Reply:
I meant like JSON values. If I have 3 stack traces and we display 15 lines before/after:

```
line 115 src/foo.js 
line 120 src/foo.js 
line 2000 src/foo.js 
```

would your `sources` contain this?

```
sources: {
  src/foo.js: { content: '<contents of lines 100 through 2016>', lineOffset: 100 }
}
```

Reply:
I think it should be

`sources: Record<string, ContentWithOffset[]>`

Note the array.

Reply:
That would enable less content. And we could deduplicate overlapping content.

Overlapping:

```
sources: {
  src/foo.js: [
    { content: '<contents of lines 100 through 131>', lineOffset: 100 }
    { content: '<contents of lines 105 through 136>', lineOffset: 105 }
    { content: '<contents of lines 1985 through 2016>', lineOffset: 1985 }        
  ]
}
```

Dedup:

```
sources: {
  src/foo.js: [
    { content: '<contents of lines 100 through 136>', lineOffset: 100 }
    { content: '<contents of lines 1985 through 2016>', lineOffset: 1985 }        
  ]
}
```

FE could count newlines in the content.

Reply:
You are both correct